### DISCUSSION TITLE
Blueprint Dynamic URL For `string` vs. `path`

### DISCUSSION THREAD

**Discussion by YTKme** [pallets/flask#5406] [Q&A]
## Context

Not sure if this is a bug or intentional feature. When using Blueprint route with Flask, I understand routes with trailing `/` can be used to handle routes with and without `/`. (e.g. `/api/do/stuff` and `/api/do/stuff/`.

When you have route:

```python
@blueprint_api.route('/project/<string:category>/', methods=['POST'])
def project_category(category):
    ...
```

When user enters a route `/project/food/`, it will return what `project_category` does and returns, but when user enters `/project/food`, it will redirect to `/project/food/`.

So if you don't want the redirect, you might do something like:

```python
@blueprint_api.route('/project/<string:category>', methods=['POST'])  # 200 OK
@blueprint_api.route('/project/<string:category>/', methods=['POST'])  # 200 OK
def project_category(category):
    ...
```

## Replicate

Now this works fine if the dynamic route uses `string` type, but it seem to be working differently for `path` type.

```python
@blueprint_api.route('/project/<path:category>', methods=['POST'])  # 308 PERMANENT REDIRECT
@blueprint_api.route('/project/<path:category>/', methods=['POST'])  # 200 OK
def project_category(category):
    ...
```

With the `path` type, even when you specify both routes, the route **without** the trailing `/` will still redirect to route with trailing `/` even though it is specified as a route.

## Expect

I was expecting the same behavior for both `string` type dynamic route parameter and `path` type dynamic route parameter?

Environment: macOS 14.3 (23D56)

- Python version: 3.10.11
- Flask version: 3.0.0

**Comment by gitsult4n** [comment]
This comes from Werkzeug routing not Flask and the redirect for `path` rules is intentional

`path` sets `part_isolating = False` because its regex matches slashes
so each `path` rule compiles into its own final regex covering the rest of the URL
Your two rules therefore become two separate branches in the router instead of sharing one state like the two `string` rules do
The trailing slash branch is tried first and raises the 308 before the other rule is ever considered

Werkzeug `2.2.3` added this behavior deliberately in its changelog

> Ensure that URL rules using path converters will redirect with strict slashes when the trailing slash is missing. :issue:`2533`

https://github.com/pallets/werkzeug/blob/main/CHANGES.rst

The fix is a single rule with `strict_slashes=False`

```python
@blueprint_api.route('/project/<path:category>/', methods=['POST'], strict_slashes=False)
def project_category(category):
    ...
```

Verified on Flask `3.1.3` with Werkzeug `3.1.8`
Both `/project/food` and `/project/food/` return 200 and `category` is `"food"` in both cases

Swapping your decorator order also avoids the redirect but then `/project/food/` matches with `category == "food/"` because `path` swallows the trailing slash
so the single rule with `strict_slashes=False` is the cleaner option
